### ISSUE TITLE
Lacking associativity of `tensor.expand` and `tensor.flatten`

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#106584]
### 🐛 Describe the bug

I guess this is another angle of https://github.com/pytorch/pytorch/issues/28090

`flatten` could realloc only the needed amount needed to deal with the required dimensions (and keep some unaffected dimensions expanded).

In this example, x1 consumes less memory than x2. If `flatten` was smarter, x2 would consume as little memory as x1.

```python
import torch
# output from meshgrid
x0 = torch.arange(100).unsqueeze(-1).expand(-1, 200)
print(x0.numel(), x0.storage().size())
#bar.py:4: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage #will be the only storage class. This should only matter to you if you are using storages directly.  To #access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
#  print(x0.numel(), x0.storage().size())

#20000 100

x1 = x0.flatten(start_dim = -2).unsqueeze(0).expand(30, -1)
print(x1.numel(), x1.storage().size())
# 600000 20000

x2 = x0.unsqueeze(0).expand(30, -1, -1).flatten(start_dim = -2)
print(x2.numel(), x2.storage().size())
# 600000 600000
```

### Versions

2.1.0 nightly

**Comment by cpuhrsch** [comment]
Does torch.compile resolve this?

**Comment by vadimkantorov** [comment]
@cpuhrsch So far I'm getting trouble from instability of selected dynamic / static kernels in https://github.com/pytorch/pytorch/issues/106614 (e.g. I want a way to force static shapes and to control which dims are static and which dynamic). Also seems currently not very good for tiling of sum-reductions...

Will update when I try this on torch.compile.

IMO a general problem is that we have no way to express that we want a copy (e.g. supporting forcing with some `copy=True` or otoh explicitly saying `copy=False` or `copy="avoid"`). Because of this, some implicit copies got depended on and now optimizing them away can be a backcompat problem.

Because of this, some application libraries like:
- xformers https://github.com/facebookresearch/xformers/blob/2d3a2217c263419243b70c53f725213d1c386b0f/xformers/ops/unbind.py#L55
- detectron2 https://github.com/facebookresearch/detectron2/blob/57bdb21249d5418c130d54e2ebdc94dda7a4c01a/detectron2/layers/wrappers.py#L50

Implement hacks to not always returning copy for unbind/cat and these cannot be upstreamed nicely.

I'm proposing more wide support for `copy="can_return_view"` or sth similar

**Comment by cpuhrsch** [comment]
@vadimkantorov - This sounds a lot like copy-on-write semantics, which we've had discussions and attempt on previously. Is that correct?

**Comment by vadimkantorov** [comment]
No, I think that it's just that flatten is doing a too strict check, and forcing a copy when it could skip it

**Comment by vadimkantorov** [comment]
Another in-the-wild hack for lack of flexible control of copy semantics for torch.cat: https://github.com/pytorch/pytorch/pull/107509
